pkg/controller/applicationlayer/applicationlayer_controller.go

```diff
@@ -0,0 +1,343 @@
+			reqLogger.Info("ApplicationLayer object not found")
+			r.status.OnCRNotFound()
+			return reconcile.Result{}, nil
+		}
+		reqLogger.Error(err, "Error querying for Application Layer")
+		r.status.SetDegraded("Error querying for Application Layer", err.Error())
+		return reconcile.Result{}, err
+	}
+
+	preDefaultPatchFrom := client.MergeFrom(applicationLayer.DeepCopy())
+
+	updateApplicationLayerWithDefaults(applicationLayer)
+
+	err = validateApplicationLayer(applicationLayer)
+
+	if err != nil {
+		reqLogger.Error(err, err.Error())
+		r.status.SetDegraded(err.Error(), "")
+		return reconcile.Result{}, nil
+	}
+
+	// Write the application layer back to the datastore, so the controllers depending on this can reconcile.
+	if err = r.client.Patch(ctx, applicationLayer, preDefaultPatchFrom); err != nil {
+		reqLogger.Error(err, "Failed to write defaults to applicationLayer")
+		r.status.SetDegraded("Failed to write defaults to applicationLayer", err.Error())
+		return reconcile.Result{}, err
+	}
+
+	variant, installation, err := utils.GetInstallation(ctx, r.client)
+
+	if err != nil {
+		if errors.IsNotFound(err) {
+			reqLogger.Error(err, "Installation not found")
+			r.status.SetDegraded("Installation not found", err.Error())
+			return reconcile.Result{}, nil
+		}
+		reqLogger.Error(err, "Error querying installation")
+		r.status.SetDegraded("Error querying installation", err.Error())
+		return reconcile.Result{}, err
+	}
+
+	managementCluster, err := utils.GetManagementCluster(ctx, r.client)
+	if err != nil {
+		reqLogger.Error(err, "Error reading ManagementCluster")
+		r.status.SetDegraded("Error reading ManagementCluster", err.Error())
+		return reconcile.Result{}, err
+	}
+
+	managementClusterConnection, err := utils.GetManagementClusterConnection(ctx, r.client)
+	if err != nil {
+		reqLogger.Error(err, "Error reading ManagementClusterConnection")
+		r.status.SetDegraded("Error reading ManagementClusterConnection", err.Error())
+		return reconcile.Result{}, err
+	}
+
+	if managementClusterConnection != nil && managementCluster != nil {
+		err = fmt.Errorf("having both a ManagementCluster and a ManagementClusterConnection is not supported")
+		reqLogger.Error(err, "")
+		r.status.SetDegraded(err.Error(), "")
+		return reconcile.Result{}, err
+	}
+
+	pullSecrets, err := utils.GetNetworkingPullSecrets(installation, r.client)
+
+	if err != nil {
+		reqLogger.Error(err, "Error retrieving pull secrets")
+		r.status.SetDegraded("Error retrieving pull secrets", err.Error())
+		return reconcile.Result{}, err
+	}
+
+	// Patch felix configuration if necessary.
+	err = r.patchFelixTproxyMode(ctx, applicationLayer)
+
+	if err != nil {
+		reqLogger.Error(err, "Error patching felix configuration")
+		r.status.SetDegraded("Error patching felix configuration", err.Error())
+		return reconcile.Result{}, err
+	}
+
+	lcSpec := applicationLayer.Spec.LogCollection
+	component := applicationlayer.ApplicationLayer(pullSecrets, installation, rmeta.OSTypeLinux,
```

Comment:
I think we should just be passing lcSpec directly into ApplicationLayer instead of passing 3 separate fields.

Reply:
@Brian-McM WDYT?
Pinging you after your comment here https://github.com/tigera/operator/pull/1505#discussion_r711209177

Reply:
I don't really have anything against passing the spec down, I just didn't think we needed another spec struct.

I also don't really have anything against pulling the fields out of the spec struct either, but maybe there should be a general config struct for all the parameters (instead of just a subset of them as there was before) in that case like we have for the some of the other components.

Reply:
after discussion on slack, we decided to pass the config as a separate struct 8ced1142ff48fde195b946f3466c202097b84902